image_proc/src/resize.cpp

```diff
@@ -0,0 +1,179 @@
+#include <rclcpp/qos.hpp>
+#include <sensor_msgs/msg/camera_info.h>
+#include <sensor_msgs/msg/image.h>
+#include <image_transport/image_transport.h>
+
+#include "resize.hpp"
+
+namespace image_proc
+{
+ResizeNode::ResizeNode(const rclcpp::NodeOptions & options) : rclcpp::Node("ResizeNode", options)
+{
+
+   auto parameter_change_cb =
+    [this](std::vector<rclcpp::Parameter> parameters) -> rcl_interfaces::msg::SetParametersResult
+    {
+      auto result = rcl_interfaces::msg::SetParametersResult();
+      result.successful = true;
+      for (auto parameter : parameters) {
+        if (parameter.get_name() == "camera_namespace") {
+          camera_namespace_ = parameter.as_string();
+          RCLCPP_INFO(get_logger(), "camera_namespace: %s ", camera_namespace_.c_str());
+          break;
+        }
+      }
+
+      image_topic_ = camera_namespace_ + "/image";
+      camera_info_topic_ = camera_namespace_ + "/camera_info";
+      connectCb();
+
+      std::lock_guard<std::mutex> lock(connect_mutex_);
+      pub_image_ = image_transport::create_publisher(this, image_topic_ + "/resize");
+      pub_info_ = this->create_publisher<sensor_msgs::msg::CameraInfo>(camera_info_topic_ + "/resize", rclcpp::QoS(1));
+
+      return result;
+    };
+
+  this->declare_parameter("camera_namespace");
+
+  interpolation_ = this->declare_parameter("interpolation", 1);
+  use_scale_ = this->declare_parameter("use_scale", true);
+  scale_height_ = this->declare_parameter("scale_height", 1.0);
+  scale_width_ = this->declare_parameter("scale_width", 1.0);
+  height_ = this->declare_parameter("height", -1);
+  width_ = this->declare_parameter("width", -1);
+
+  rclcpp::Parameter parameter;
+  if (rclcpp::PARAMETER_NOT_SET != this->get_parameter("camera_namespace", parameter)) {
+    std::cout << "Camera space set" << std::endl;
+    parameter_change_cb(this->get_parameters({"camera_namespace"}));
+  }
+
+  this->set_on_parameters_set_callback(parameter_change_cb);
+};
+
+// Handles (un)subscribing when clients (un)subscribe
+void ResizeNode::connectCb()
+{
+  std::lock_guard<std::mutex> lock(connect_mutex_);
+  
+  if (!sub_image_)
+  {
+    std::cout << "subscribe sub_image_" << std::endl;
+    sub_image_ = image_transport::create_subscription(this, image_topic_, std::bind(&ResizeNode::imageCb, this, std::placeholders::_1), "raw");
+  }
+  
+  if (!sub_info_)
+  {
+    std::cout << "subscribe sub_info_" << std::endl;
+    sub_info_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(camera_info_topic_, rclcpp::QoS(1), std::bind(&ResizeNode::infoCb, this, std::placeholders::_1));
+  }
+}
+
+void ResizeNode::infoCb(const sensor_msgs::msg::CameraInfo::SharedPtr info_msg)
+{
+  sensor_msgs::msg::CameraInfo::SharedPtr dst_info_msg = info_msg;
+  
+  double scale_y;
+  double scale_x;
+  if (use_scale_)
+  {
+    scale_y = scale_height_;
+    scale_x = scale_width_;
+    dst_info_msg->height = static_cast<int>(info_msg->height * scale_height_);
```

Comment:
I appreciate your care with casting

Reply:
Thanks, but can't take credit for that, was in the ROS1 code. :)